### ISSUE TITLE
Separate persistent data based on variant aircraft needs

### ISSUE THREAD

**Issue by wlbragg** [c172p-team/c172p#1421]
This issue is to complete the work of separating all persistent data by variant needs.

A complete review of persistent data per variant aircraft is required.

One of these issues is the transponder, apparently the fg1000 possibly uses less properties than the default 172s which is causing issues with persistent data saves and loads.

[dany93](https://github.com/dany93) commented [yesterday](https://github.com/c172p-team/c172p/issues/1410#issuecomment-1625082311) • 
Commenting out the knob-mode and ident-btn is enough to make it work several times without deleting the cache.
Perfectly consistent with my https://github.com/c172p-team/c172p/issues/1410#issuecomment-1424251304 (in the cache).

The transponder digits are saved and reloaded in FG1000.

~~~
# DME saves power-btn in ki266.nas
# ADF saves its properties in ki87.nas
# TRANSPONDER (KT76A)
aircraft.data.add(
#    "instrumentation/transponder/inputs/knob-mode",
#    "instrumentation/transponder/inputs/ident-btn",
    "instrumentation/transponder/inputs/digit[0]",
    "instrumentation/transponder/inputs/digit[1]",
    "instrumentation/transponder/inputs/digit[2]",
    "instrumentation/transponder/inputs/digit[3]",
    "instrumentation/transponder/inputs/dimming-norm",
);
The c182t (FG1000) in aircraft-data, which works with no such issue, doesn't have these lines saved.
~~~

**Comment by wlbragg** [comment]
@dany93 would you mind trying to manually changing the properties from

"instrumentation/transponder/inputs/knob-mode" = 4
 "instrumentation/transponder/inputs/ident-btn' = false
to 
"instrumentation/transponder/inputs/knob-mode" = 3
 "instrumentation/transponder/inputs/ident-btn" false = true

Then exiting the sim properly and trying to restart, just curious if this fixes it for future startups?

**Comment by dany93** [comment]
No. Same issue,the PFD has no animations.
with aircraft-data/c172p-fg1000-kap.xml:
```
    <transponder>
      <inputs>
        <knob-mode type="int">3</knob-mode>
        <ident-btn type="bool">true</ident-btn>
        <digit type="int">0</digit>
        <digit n="1" type="int">0</digit>
        <digit n="2" type="int">2</digit>
        <digit n="3" type="int">1</digit>
      </inputs>
```

**Comment by dany93** [comment]
>the fg1000 possibly uses less properties than the default 172s 

Less properties or not the same? Different names?
knob-mode number and ident-btn exist and can be changed in the FG1000 variant properties. Changeable by clicking on the XPDR buttons.

**Comment by wlbragg** [comment]
I've scoured these properties and I am at a loss as to why your system is having issues. They are the same properties. I thought maybe their properties were a different "type" int VS bool. But I can't find that either. We don't have to make them persistent, but I can't figure out why in your case they can't be but in mine they are fine. What happens if after you have your new version up and running and then add code back in to cache those two properties, does it then stop working again? What about without this new PR we just merged? Does this mean it is definitely something in the "https://github.com/c172p-team/c172p/issues/1410" PR that causes the failure? This PR has nothing to do with persistent data. So do other branches have this issue?

**Comment by dany93** [comment]
Tested **in master** after git pull.
I noticed that the two "bad" lines still are in avionics.nas. then, consequently, in the cache file. No change.
In master, as in Issue-1410 branch, I had to manually delete the two lines in both files to do these tests.
>What happens if after you have your new version up and running and then add code back in to cache those two properties, does it then stop working again?

As expected, the PFD stops working again.
>What about without this new PR we just merged?

What new PR? I don't see what it changes.
For me, the last PR is `2364c1e28`
```
    Merge pull request #1420 from c172p-team/revert-1412-Issue-1410
    
    Revert "Material animation and effects definitions come after knob an…
```
>Does this mean it is definitely something in the "https://github.com/c172p-team/c172p/issues/1410" PR that causes the failure? This PR has nothing to do with persistent data. So do other branches have this issue?

I've made a test in master with `0f8e4e5e3`
```
Merge pull request #1416 from c172p-team/Issue-1414
    
    FDM: decrease pitch up tendency at full thrust.
```
Same symptom, the PFD does not work
>I can't figure out why in your case they can't be but in mine they are fine

I totally agree.

**Comment by dany93** [comment]
About the persistent properties:
- `instrumentation/transponder/inputs/knob-mode` can be useful. Although not essential.
- `instrumentation/transponder/inputs/ident-btn`  doesn't make sense according to my understanding of transponder (no real practice). I think the "ident" signal should be temporary, short, at the controller's request.

**Comment by dany93** [comment]
Thoughts:
Separating all persistent data "by variant needs" is probably a hard work.
- The easy solution (not perfect) is deleting the two properties in avionics.nas. So that they are not saved in the cache at Exit.
- We can assume that I'm the only one to get the issue. In which case we let the code unchanged, waiting for someone else to complain.
- It might also be a bug in FG1000. In which case you risk doing a hard work to circumvent an error, not to build the solution.

**Comment by wlbragg** [comment]
>Does this mean it is definitely something in the "https://github.com/c172p-team/c172p/issues/1410" PR that causes the failure?

What I am wondering is if you have this issue in any version that contains the fg1000? Like any versions before [this issue](https://github.com/c172p-team/c172p/pull/1412) that you first noticed it. I guess basically I am wondering if you bisect versions that contain fg1000 if they all have this issue for you. Can you identify when this started, was it the very first version that contained an fg1000 option. Note: there were also versions where it was a part of the base aircraft selectable from the GUI and not a hard variant, meaning it wasn't a stand alone variant. Did any of those cause this issue?

I think just testing [2020.4](https://github.com/c172p-team/c172p/releases/tag/version%2F2020.4) would answer my above question. It might be difficult to find any specific versions where it was all combined and not a standalone variant. They should all be after 2020.3 up to 2020.4. In fact at one point the fg1000 even shared the same electrical system and the rest of the variants. 

If this is too much trouble to look for, we can either ignore and see if others complain or modify the cached data.

I think I will still likely have at least one commit attached to this issue to finish cleaning up the separation of variant data be it cached data or just properties or systems that the fg1000 doesn't use. It was mostly done in prior work. I just need to go through it one more time and verify that all properties, systems and data not needed in the fg1000 are removed.

**Comment by dany93** [comment]
I've done a kind of git bisect (not exactly exhaustive, but not far).
I found no version without the PFD animation issue.

From the older to the most recent.

f74cf868c4
Begin fg1000 variant
2020-10-13
I couldn't get avionics "ON".
______________________________

da26466e8
FG1000 variant progress
2020-10-27
BUG
_________________________________
799db0808
Breaker work, use GFC700 auto pilot
2020-10-29
BUG
_______________________________
c8ddf8229
FG1000 autopilot variant option, amphibious gear ctrl configuration
2020-10-30
BUG
____________________________
90ed032fc
 FG1000 electrical system
2020-11-03
BUG
__________________________________
abd2f82bcf9
FG1000 variant needs complete isolation from other variants
2021-02-10
BUG
____________________________
f0e108bca8
tag version/2020.4
2021-08-22
BUG
_________________________

Among those tested, I have the issue ("BUG") on every version on which the FG1000 could be switched ON.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Maybe Important:
**I do not have the issue with the J3Cub-FG1000-set.**
It starts (3 times, the first one with no cache) even after "Quit" with no issue.

With **cache** (even after two "Quit"):
aircraft-data/J3Cub-FG1000.xml:
```
    <transponder>
      <inputs>
        <knob-mode type="int">4</knob-mode>
        <ident-btn type="bool">false</ident-btn>
        <digit type="int">0</digit>
        <digit n="1" type="int">0</digit>
        <digit n="2" type="int">2</digit>
        <digit n="3" type="int">1</digit>
      </inputs>
    </transponder>
```

J3Cub **avionics**.nas:
```
# TRANSPONDER (KT76A)
aircraft.data.add(
    "instrumentation/transponder/inputs/knob-mode",
    "instrumentation/transponder/inputs/ident-btn",
    "instrumentation/transponder/inputs/digit[0]",
    "instrumentation/transponder/inputs/digit[1]",
    "instrumentation/transponder/inputs/digit[2]",
    "instrumentation/transponder/inputs/digit[3]",
    "instrumentation/transponder/inputs/dimming-norm",
);
```
Same lines as c172.

**Comment by dany93** [comment]
(Reminder)
When I have the PFD failing, I see these lines in the log:
```
         [INFO]:nasal      #  FG1000  #
(...)

   13.99 [INFO]:nasal      Traffic pins off
   14.02 [INFO]:nasal      string index 3 out of bounds (size: 3)
   14.02 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstruments.nas
   14.02 [INFO]:nasal      956
   14.02 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstrumentsController.nas
   14.02 [INFO]:nasal      478
   14.02 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstrumentsController.nas
   14.02 [INFO]:nasal      540
   14.02 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Nasal/emesary.nas
   14.02 [INFO]:nasal      179
   14.02 [ALRT]:nasal      Recipient PFDInstrumentsController_PFD has been removed from transmitter (GlobalTransmitter) because of the above error
   14.02 [INFO]:nasal      Annunciator panel: initialized
```

Instead of these two with nothing between (when it works)
```
   14.05 [INFO]:nasal      Traffic pins off
   14.05 [INFO]:nasal      Annunciator panel: initialized
```
Can it be a clue? Do you understand this?

**Comment by wlbragg** [comment]
>Can it be a clue? Do you understand this?

Yeah, possibly. Thank you so much for documenting this.

"The string index 3 out of bounds" is one I have seen before. I think it is the cause of your issue. I thought that type of error has always been associated with a GUI error and very hard to track down, if this is what I am thinking it is. I've had it show before with addon stuff and no debugging info in a menu to track it. May not be the case here, but definitely a clue. Thanks, I need to look at this closely. 

Recipient PFDInstrumentsController_PFD has been removed from transmitter (GlobalTransmitter) because of the above error
   14.02 [INFO]:nasal      Annunciator panel: initialized

That is the effect of the error and the reason the PFD fails.

The fg1000 working in the Cub is another good clue as it will give me some comparisons, GUI code and fg1000 nasal setup. Good info, again, thank you!

**Comment by wlbragg** [comment]
FYI:

>string index 3 out of bounds (size:  3)

If I had to guess I would say that is a string field with a set size being fed data that exceeds that size. I think I usually see that when a menu property is fed bad or incorrect data. But I don't know why your getting it, especially if your not doing any GUI at the time of startup. But maybe the fg1000 code is reading persistent data at startup that is not of the correct type or size?

**Comment by wkitty42** [comment]
On 7/10/23 1:17 PM, wlbragg wrote:
> FYI:
> 
>     string index 3 out of bounds (size: 3)
> 
> If I had to guess I would say that is a string field with a set size being fed 
> data that exceeds that size. I think I usually see that when a menu property is 
> fed bad or incorrect data. But I don't know why your getting it, especially if 
> your not doing any GUI at the time of startup. But maybe the fg1000 code is 
> reading persistent data at startup that is not of the correct type or size?

i'm thinking of an array here... like there's some array of strings(?) and 
there's currently only 3 entries in it, 0-2... then something is trying to 
access array[4] which does not (yet?) exist... maybe?

array[0]="foo"
array[1]="bar"
array[2]="fubar"
array[3]="snafu"

perhaps it is an "off by one" defect where the counter causing this to appear 
needs to be (eg) foo-1 to fit within the array bounds...

**Comment by dany93** [comment]
I also had this idea. But cautiously, knowing my poor skills in this code.

In flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/**PFDInstruments.nas**,
line 67
```
    obj.addTextElements([
         (skip...)
      "XPDR-DIGIT-3-text", "XPDR-DIGIT-2-text", "XPDR-DIGIT-1-text", "XPDR-DIGIT-0-text",

```
would define a 4-elements array. Enough for the XPDR code.

And lines 946 - 957
```
    if (ident) {
          (skip...)
    var codes = split("", code);
    me.getElement("XPDR-DIGIT-3-text").setText(chr(code[0])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-2-text").setText(chr(code[1])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-1-text").setText(chr(code[2])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-0-text").setText(chr(code[3])).setColor(r,g,b);
  },
Should work... ?
```

**Comment by dany93** [comment]
From bad to worse...

Yesterday, I had the **J3Cub-FG1000**-set working. https://github.com/c172p-team/c172p/issues/1421#issuecomment-1628593370
This day, **the autostart doesn't work** (no longer works), hence the FG1000 doesn't switch ON. Nasal errors. Even after deleting the aircraft cache. I understand nothing... 
```
  169.22 [ALRT]:nasal      Nasal runtime error: undefined symbol: electrical
  169.22 [ALRT]:nasal        at /home/dany/FG/github/J3Cub/Nasal/j3cub.nas, line 24
  169.22 [ALRT]:nasal        called from: /sim/bindings/menu/binding[126], line 1
```
This nasal code is really a calamity...

**Comment by dany93** [comment]
**J3Cub FG1000: Fixed after a git pull**.
Which is particularly weird, I had changed nothing on the J3Cub. I had checked that my local was up to date with my local origin master.

But now, it is like the c172p. **The J3Cub FG1000 works only once after having deleted the aircraft cache**.
More consistent, but it doesn't help.

**Comment by wlbragg** [comment]
Wow, one is ESI and one is PFD. This is interesting. Good you documented steps (or lack of) that caused this. I'll have to trace through that part of the fg1000 code which I am not strong with. It was all a little abstract to me when I added the fg1000 code.

>Should work... ?
var codes = split("", code);

Without going through the code that builds the variable "code", yes, on the surface that is good. The issue could be the manipulation of the variable "code" to insure that it is a 4 digit number, as per what the comments say...

# Ensure the code is a 4 digit string representation of a number.
# Normally this means padding with 0's at the left e.g.  42 becomes 0042.
# In editing mode we enter values from the left, so we want " " padding on the
# right, so 42 becomes "42  ".  Note also that the code is a string value
# so we can have "004 "

That is exactly the place this type of error could creep in, not saying it is though.

**Comment by wlbragg** [comment]
That's weird, I read that wrong, The ESI is being shut down is when it works, then it is the PFD when it doesn't. Can you verify after you remove the cache and start successfully (the J3Cub) that the ESI data is working? That is the right glass engine performance data, on the left side of the right glass.

I see it
No such member: EBusVolts
I need to see what happened there. I know what that is.

**Comment by wlbragg** [comment]
@dany93 is there anyway you can identify the new commits you brought in when you updated that broke the j3cub?

That might be really helpful.

**Comment by wlbragg** [comment]
Well, thank you for finding the bug with the Piper. The ESI is not working in either the J3 or the PA-18. At least I know what that is.
I find it interesting that this has likely been broken for awhile yet you are the first to notice it. Just goes to show you how either, few people test or fly these aircraft, or how seldom they upgrade when new work is done. The fact that I didn't notice it says something as well.

**Comment by dany93** [comment]
> is there anyway you can identify the new commits you brought in when you https://github.com/c172p-team/c172p/issues/1421#issuecomment-1630634071?

Not easy, not certain, because I'm not sure of the last commit it was at yesterday when I had it working. I don't understand why I had it working yesterday and why it didn't work this morning. I did no git pull yesterday.
I noted that I did a "git pull" on 05 April 2022. The might be my last one.
That would mean that I was at 55624bef, " Interior/Exterior temperature gauge correction" 2021-12-23.
This morning, by "git pull", I would have downloaded between 85d2b4569 " Trim should allow for positive and negative adjustment" 
2022-11-02 and 7a2fdb315b6 "Parking brake, maintenance" 2023-06-21.

**Comment by wlbragg** [comment]
OK, thanks. I'll see if what you gave me helps.
FYI: the J3Cub ESI should now work in all repositories, SVN, GitHub and GitHub-Datcom versions. I'm curious if that maybe fixed the PFD cache issue in the Cub for you, I doubt it!

**Comment by dany93** [comment]
git pull, **5a6ff0d66e "Remove unused file, formatting" 2023-07-11**.

After deleting the aircraft cache, as usually.

- First launching:
Everything works, PFD and ESI (RPM, Oil,..). No noticeable nasal error message in the FG1000 log.

- Second one (with the aircraft cache):
ESI works, PFD doesn't. No surprise.
Usual error messages in the log:
```
   12.96 [INFO]:nasal      string index 3 out of bounds (size: 3)
   12.96 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstruments.nas
   12.96 [INFO]:nasal      956
   12.96 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstrumentsController.nas
   12.96 [INFO]:nasal      478
   12.96 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstrumentsController.nas
   12.96 [INFO]:nasal      540
   12.96 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Nasal/emesary.nas
   12.96 [INFO]:nasal      179
   12.96 [ALRT]:nasal      Recipient PFDInstrumentsController_PFD has been removed from transmitter (GlobalTransmitter) because of the above error
```

**Comment by dany93** [comment]
I confirm:
With **55624bef**, " Interior/Exterior temperature gauge correction" 2021-12-23, **the engine does not start** (at least by autostart).
(with the cache deleted)
```
   11.75 [INFO]:nasal      No such member: EBusVolts
   11.75 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal//EIS/EIS-C182T.nas
   11.75 [INFO]:nasal      53
   11.75 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal//EIS/EISController.nas
   11.75 [INFO]:nasal      58
   11.75 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal//EIS/EISController.nas
   11.75 [INFO]:nasal      89
   11.75 [INFO]:nasal      /home/dany/fgfs-B/install/flightgear/fgdata/Nasal/emesary.nas
   11.75 [INFO]:nasal      179
   11.75 [ALRT]:nasal      Recipient EISController_MFD has been removed from transmitter (GlobalTransmitter) because of the above error
```

```
   36.45 [ALRT]:nasal      Nasal runtime error: undefined symbol: electrical
   36.45 [ALRT]:nasal        at /home/dany/FG/github/J3Cub/Nasal/j3cub.nas, line 24
   36.45 [ALRT]:nasal        called from: /sim/bindings/menu/binding[126], line 1
```

**Comment by dany93** [comment]
Knowing that the c182t PFD works, I've made a test.

In the 182t cache .fgfs/aircraft-data/c182t.xml, I've replaced the initial lines (530 - 532)
```
    <transponder>
      <id-code type="int">5421</id-code>
    </transponder>
```
for these ones, similar to those in aircraft-data/c172p.xml.
```
    <transponder>
      <inputs>
        <knob-mode type="int">5</knob-mode>
        <digit type="int">3</digit>
        <digit n="1" type="int">0</digit>
        <digit n="2" type="int">2</digit>
        <digit n="3" type="int">1</digit>
      </inputs>
    </transponder>
```
In the c182t, the PFD works, not only the first time.
If I change the XPDR code by the buttons in the FG1000 XPDR, this code is saved and retrieved at the next launching.
Not the knob-mode, which is normal (not in the persistent list).

This means that, in the c182, the FG1000 can handle these lines as they are written in the c172 cache. The problem lies with c172 (and J3Cub).

The way to change the cache should not matter, only the file content matters at launching.
In the c182, this cache is saved by xml (in c182t-set.xml). Differently from the c172p.

Do the c172 and c182 use the same FG1000?
Do the c172 has nasal lines added for the FG1000 persistent input?

One more not the solution, by in the hope it can help to converge....

**Comment by wlbragg** [comment]
>Do the c172 and c182 use the same FG1000?

Kind of. There is a slight difference with the ESI, I think.

>This means that, in the c182, the FG1000 can handle these lines as they are written in the c172 cache. The problem lies with c172 (and J3Cub).

Great clues, I actually used the c182 version when I corrected the J3Cub to refresh my memory of all the parts. Then I changed it back to my version once I got all the connections sorted out. I can easily swap out everything in the c172p to use the same exact configuration of the c182 (but still using the persistent data) and we can see if that fixes it. I don't have any problem doing that as they are basically the same data. Originally the only difference was the ESI, I added the EBus (it is the emergency electrical bus on the c172sp) to the ESI portion locally as Stuart hadn't added it in fgdata at that time. I'm not sure he added it yet.

I could try in order...
1) Try using the c182 PFD implementation and the c172sp ESI implementation.
2) Try using the c182 PFD implementation and the c182 ESI implementation.

See if that magically corrects the cache issue?

**Comment by wlbragg** [comment]
My c182t 79b1eb3eefa78aa2a5eaa254011e92aac71671fd from https://github.com/HHS81/c182s the ESI is working.
My c182t loaded from fgaddon/qtlauncher is also working.

**Comment by dany93** [comment]
Right. 
- By locally creating and cloning a c182s github repository from HHS81, 79b1eb3eefa, the EIS works too.
- Also by updating my copy from FGAddon.
==> the c182s EIS works.

Sorry.

**Comment by dany93** [comment]
I confirm [the test above](https://github.com/c172p-team/c172p/issues/1421#issuecomment-1634074115).
The c182t PFD still works by replacing the initial lines in the cache for these from the c172 as written above.
The EIS works too.

**Comment by dany93** [comment]
(--aircraft=c172p-fg1000-kap)

The persistent `<knob-mode...>` property is clearly the source of the PFD failure. It can give it by itself.

  In the cache file c172p-fg1000-kap.xml, **if I delete all the lines, leaving only these**, as (below):
```
<?xml version="1.0"?>

<PropertyList>
  <instrumentation>
    <transponder>
      <inputs>
        <knob-mode type="int">5</knob-mode>
      </inputs>
    </transponder>
  </instrumentation>
</PropertyList>
```
The PFD animation doesn't work, and the usual PFD error lines are displayed in the log.

Also, surprisingly, the "knob-mode" property is correctly passed to the XPDR (here: 5 = ALT).

**Comment by wlbragg** [comment]
The error points to "me.getElement("XPDR-DIGIT-0-text").setText(chr(code[3])).setColor(r,g,b);"

per... 

/home/dany/fgfs-B/install/flightgear/fgdata/Aircraft/Instruments-3d/FG1000/Nasal/MFDPages/PFDInstruments/PFDInstruments.nas
   12.96 [INFO]:nasal      956

I think the error comes in when NOT ident
~~~
if (ident) {
      me.getElement("XPDR-MODE-text").setText("IDNT").setColor(r,g,b);
    } else {
      me.getElement("XPDR-MODE-text").setText(TRANSPONDER_MODES[mode]).setColor(r,g,b);
    }

    var codes = split("", code);
    me.getElement("XPDR-DIGIT-3-text").setText(chr(code[0])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-2-text").setText(chr(code[1])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-1-text").setText(chr(code[2])).setColor(r,g,b);
    me.getElement("XPDR-DIGIT-0-text").setText(chr(code[3])).setColor(r,g,b); <-line 956
~~~

I think TRANSPONDER_MODES[mode] somehow might become a 3 char instead of the needed 4 char?

I haven't spent enough time looking through the rest of the code in PFDInstrumentsController.nas and deciphering it. It's hard to follow. Pretty much guessing with a little logic attached.

I'm wondering if changing the transponder mode and then exiting to save different persistent data (try exiting using all different possible mode choices) might have any effect?

Or trying the same thing using different transponder ident numbers, don't use leading 0's or maybe no 0's at all and exiting to change and test different ident numbers, trying to not cause the code to have to add to a less than 4 digit transponder number?

**Comment by wlbragg** [comment]
Actually, after looking closer I discarded TRANSPONDER_MODES[mode] being used by chr(code[3]. chr(code[3] comes in in the PFDInstrumentsController.nas code. Again I haven't dissected that code to try and trace exactly what all is involved with the array code[].

**Comment by wlbragg** [comment]
@dany93 
I just managed to duplicate your error. Can you post the contents of the following files?
Nasal\Interfaces\SelectableInterfaceController.nas
The top 30 lines of Nasal\electrical-fg1000.nas
Do you have the following files...
Nasal\Interfaces\SelectableFuelInterface.nas
Nasal\Interfaces\SelectableFuelPublisher.nas

**Comment by wlbragg** [comment]
I don't believe this, it's now broke just like yours in both the c172 and the J3/Pa-18. I reset everything back to where it was working and it still doesn't  This is truly the weirdest thing I have ever seen.

**Comment by wlbragg** [comment]
I'm curious, where and how does the marker beacon work in the fg1000, is there one?

**Comment by wlbragg** [comment]
Observations...

Mine is both knob-mode and ident properties that are at issue.

The error string index 3 out of bounds (size: 3) could be referring to the input into 
updateTransponder(me._transponder_mode, me._transponder_code, me._transponder_ident);

Expecting three strings or args, but not getting that?

Ident is the 3rd string or arg in the function and is the one that can be truncated and is typed checked to make sure it conforms.

I also noticed none of the instrument\transponder\??? properties are initialized anywhere in any set file. I added them but it didn't seem to cure anything. I thought that it was a possibility. I have seen some very strange stuff when a property isn't initialized correctly.

**Comment by dany93** [comment]
>I just managed to duplicate your error.

It should help to research and fix the issue.
I agree, really weird....
 >Can you post the contents of the following files?

Nasal\Interfaces\SelectableInterfaceController.nas

```
var nasal_dir = getprop("/sim/fg-root") ~ "/Aircraft/Instruments-3d/FG1000/Nasal/";
var aircraft_dir = getprop("/sim/aircraft-dir");
io.load_nasal(nasal_dir ~ 'Interfaces/PropertyPublisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/PropertyUpdater.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/NavDataInterface.nas', "fg1000");
io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableEISPublisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericNavComPublisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericNavComUpdater.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericFMSPublisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericFMSUpdater.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericADCPublisher.nas', "fg1000");
io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableFuelInterface.nas', "fg1000");
io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableFuelPublisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GFC700Publisher.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GFC700Interface.nas', "fg1000");

var SelectableInterfaceController = {

  _instance : nil,

  INTERFACE_LIST : [
    "NavDataInterface",
    "SelectableEISPublisher",
    "GenericNavComPublisher",
    "GenericNavComUpdater",
    "GenericFMSPublisher",
    "GenericFMSUpdater",
    "GenericADCPublisher",
    "SelectableFuelInterface",
    "SelectableFuelPublisher",
    "GFC700Publisher",
    "GFC700Interface",
  ],

  # Factory method
  getOrCreateInstance : func() {
    if (SelectableInterfaceController._instance == nil) {
      SelectableInterfaceController._instance = SelectableInterfaceController.new();
    }

    return SelectableInterfaceController._instance;
  },

  new : func() {
    var obj = {
      parents : [SelectableInterfaceController],
      running : 0,
    };

    return obj;
  },

  start : func() {
    if (me.running) return;

    foreach (var interface; SelectableInterfaceController.INTERFACE_LIST) {
      var code = sprintf("me.%sInstance = fg1000.%s.new();", interface, interface);
      var instantiate = compile(code);
      instantiate();
    }

    foreach (var interface; SelectableInterfaceController.INTERFACE_LIST) {
      var code = 'me.' ~ interface ~ 'Instance.start();';
      var start_interface = compile(code);
      start_interface();
    }

    me.running = 1;
  },

  stop : func() {
    if (me.running == 0) return;

    foreach (var interface; SelectableInterfaceController.INTERFACE_LIST) {
      var code = 'me.' ~ interface ~ 'Instance.stop();';
      var stop_interface = compile(code);
      stop_interface();
    }
  },

  restart : func() {
    me.stop();
    me.start();
  }
};

```

>The top 30 lines of Nasal\electrical-fg1000.nas
```

var nasal_dir = getprop("/sim/fg-root") ~ "/Aircraft/Instruments-3d/FG1000/Nasal/";
io.load_nasal(nasal_dir ~ 'FG1000.nas', "fg1000");
var aircraft_dir = getprop("/sim/aircraft-dir");
io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableInterfaceController.nas', "fg1000");

var interfaceController = fg1000.SelectableInterfaceController.getOrCreateInstance();
interfaceController.start();

# Create the FG1000
var fg1000system = fg1000.FG1000.getOrCreateInstance();

# Create a PFD as device 1, MFD as device 2
fg1000system.addPFD(1);
fg1000system.addMFD(2);

# Display the devices
fg1000system.display(1);
fg1000system.display(2);

#  Display a GUI version of device 1 at 50% scale.
#fg1000system.displayGUI(1, 0.5);

##
# Procedural model of a Cessna 172S electrical system.  Includes a
# preliminary battery charge/discharge model and realistic ammeter
# gauge modeling.
#

##
# Beacon and Strobe intermittent instantiation
#
```


>Do you have the following files...

Nasal\Interfaces\SelectableFuelInterface.nas
Yes. (9 jul 2023)

Nasal\Interfaces\SelectableFuelPublisher.nas
Yes. (9 jul 2023)

**Comment by dany93** [comment]
>I'm curious, where and how does the marker beacon work in the fg1000, is there one?

I don't know. I haven't seen it working.
However, I haven't a large and good practice of the FG1000. And, even when I see it working in the "standard" instrument configuration, I must admit that I do not pay much attention to it...

Let me know if you'd like me to look into it further.

**Comment by wlbragg** [comment]
>Is "Ident" which is truncated? I'd rather understand the XPDR "code" (4 digits).

Yes, XPNDR, sorry.

It's been awhile since I used the radios but it is the localizer freq that activates the ILS (NAV) and I think is needed for the markers to activate. The marker and ILS is working at [KICT](https://flightaware.com/resources/airport/KICT/IAP/ILS+OR+LOC+RWY+01L) 109.1 NAV.

So at least the radio features that work in the default aircraft work in the fg1000. They are all controlled using the audio panel which is also working. I couldn't remember exactly what all I had activated in the audio panel, apparently most of it.

I still can't believe I ended up with the bug! All I did was change to the GenericInterfaceController instead of our custom SelectableInterfaceController and then back. After I changed it back to the SelectableInterfaceController it had the bug. I even created a new local branch to do this. After I got the bug I tried to switch back to the master branch that I didn't have the bug in and never changed a thing and I still had this bug. It's a quandary for sure?

My last test before just eliminating the persistent lines that are at issue is to try both the generic and c182t InterfaceControllers. Quit honestly I forgot the outcome of the last test because I was so stunned by the fact that I acquired the bug.

**Comment by wlbragg** [comment]
OK, changing the following in c172p\Nasal\electrical-fg1000.nas to "All Generic FG1000"
~~~
var nasal_dir = getprop("/sim/fg-root") ~ "/Aircraft/Instruments-3d/FG1000/Nasal/";
io.load_nasal(nasal_dir ~ 'FG1000.nas', "fg1000");
var aircraft_dir = getprop("/sim/aircraft-dir");
#io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableInterfaceController.nas', "fg1000");
io.load_nasal(nasal_dir ~ 'Interfaces/GenericInterfaceController.nas', "fg1000");

#var interfaceController = fg1000.SelectableInterfaceController.getOrCreateInstance();
var interfaceController = fg1000.GenericInterfaceController.getOrCreateInstance();
interfaceController.start();
~~~
**Cache is not a problem.**

Or using the c182t custom files
Changing the following in c172p\Nasal\electrical-fg1000.nas
~~~
var nasal_dir = getprop("/sim/fg-root") ~ "/Aircraft/Instruments-3d/FG1000/Nasal/";
io.load_nasal(nasal_dir ~ 'FG1000.nas', "fg1000");
#var aircraft_dir = getprop("/sim/aircraft-dir");
#io.load_nasal(aircraft_dir ~ '/Nasal/Interfaces/SelectableInterfaceController.nas', "fg1000");

#point at your local c182t files
io.load_nasal('G:/Aircraft/Development-Aircraft/c182s/Nasal/c182t-InterfaceController.nas', "fg1000");  # use custom controller
var interfaceController = fg1000.GenericInterfaceController.getOrCreateInstance();

#var interfaceController = fg1000.SelectableInterfaceController.getOrCreateInstance();
interfaceController.start();
~~~
and modifying the EIS call in G:\Aircraft\Development-Aircraft\c182s\Nasal\c182t-InterfaceController.nas to ...
~~~
#io.load_nasal(aircraft_dir ~ '/Nasal/c182t-EISPublisher.nas', "fg1000");
#point at your local c182t files
io.load_nasal('G:/Aircraft/Development-Aircraft/c182s/Nasal/c182t-EISPublisher.nas', "fg1000");
~~~
**Cache is not a problem.**

Put everything back to what is currently in the master branch and guess what?
**Cache is not a problem.**
The bug is gone again.

Maybe if you get the time you can go through at least the first test of changing to the "All Generic FG1000" starting without clearing the cache a couple times and then changing back to what you have currently or what is in master and see if you also lose the bug.

**Comment by wlbragg** [comment]
The J3/PA-18 is now working without the bug. No changes at all to the files there. I don't understand how this is even possible. I want to say maybe it never really failed, even though I know better.

**Comment by wlbragg** [comment]
@dany93 I know I am kind of jumping around here. But does this look like this is combining the engine hours of the 180hp and 160hp engines instead of saving their individual usage? Would you know why, don't you think these should be saved and displayed per their individual usage?

**Comment by wlbragg** [comment]
Three is also a filter using the total hours as well. I don't understand why we are combining the different engine use hours?
~~~
    <filter>
        <name>Engine Total Hobbs Hours</name>
        <type>gain</type>
        <input>
            <expression>
                <div>
                    <sum>
                        <property>/sim/time/hobbs/engine[0]</property>
                        <property>/sim/time/hobbs/engine[1]</property>
                    </sum>
                    <value>3600.0</value>
                </div>
            </expression>
        </input>
        <output>
            <property>/instrumentation/clock/hobbs-meter-hours</property>
        </output>
    </filter>
~~~

**Comment by dany93** [comment]
>if you get the time you can go through at least the first test of changing to the "All Generic FG1000" starting without clearing the cache a couple times and then changing back to what you have currently or what is in master and see if you also lose the bug.

I did this test (with `GenericInterfaceController.na`s instead of `SelectableInterfaceController.nas`).
- Prior to first launch, my cache had been cleaned from the (bad) `knob-mode` and `indent-btn` lines.
The PFD worked, as expected.
- next couple of times, the PDF did not work. The bug was still here, with the usual nasal error lines in the log.
No fix.

**Comment by dany93** [comment]
> I know I am kind of jumping around here. But does this look like this is combining the engine hours of the 180hp and 160hp engines instead of saving their individual usage? Would you know why, don't you think these should be saved and displayed per their individual usage?

I didn't write these lines but I agree, that would be logical. This is not the same aircraft.

However, I have the impression that almost nobody looks at this Hobbs meter values. Is it worth making the files more complicated?

As you wrote, no connection with out current PFD issue.

**Comment by wkitty42** [comment]
On 7/18/23 9:41 AM, dany93 wrote:
>     I know I am kind of jumping around here. But does this look like this is
>     combining the engine hours of the 180hp and 160hp engines instead of saving
>     their individual usage? Would you know why, don't you think these should be
>     saved and displayed per their individual usage?
> 
> I didn't write these lines but I agree, that would be logical. This is not the 
> same aircraft.

i tend to agree...

> However, I have the impression that almost nobody looks at this Hobbs meter 
> values. Is it worth making the files more complicated?

when i flew the c172p more often, i kept up with it... even going so far as to 
manually copy the value to put in the new aircraft-data file when i cleared it 
for some reason...

i didn't really think about the different configurations being different craft 
but it makes perfect sense... if we had them as variants via separate -set files 
instead of internal configuration switching, it might be easier to keep them 
separated... idk... my code reading (and coding) isn't what it used to be...

**Comment by wlbragg** [comment]
>if we had them as variants via separate -set files
instead of internal configuration switching, it might be easier to keep them
separated

Well we kind of do, When I set up the existing variants I made distinct combinations like the amphibious and special undercarriages used the 180 engine because it was most likely to be that configuration IRL. Where the standard and bush gear use the 160. So there is that. The only thing I didn't make was variants for each engine type combined with each undercarriage. If we made all those startup variants, then you wouldn't absolutely have to have the in-sim change available. But it is nice to be able to switch them on the fly.

>Is it worth making the files more complicated?
It's only a few lines of code. I went ahead and did it for the Cub/PA-18. I'm going to make another issue and copy this conversation over to it and make the change. I was quite surprised it hadn't already been done.

**Comment by wlbragg** [comment]
[See new issue for Hobbs meter.](https://github.com/c172p-team/c172p/issues/1424)

@dany93 I can't explain why this bizarre behaviour. I'm not done testing this. The fact that I was able to recreate and then correct your issue locally is something to pursue, I guess.

**Comment by wlbragg** [comment]
@dany93  can you check the J3Cub FG1000 variant, not the PA-18 Supercub variant. My PFD is working on the J3 but not the PA-18. I hadn't noticed that until now.

**Comment by dany93** [comment]
On both, the J3Cub FG1000 variant and the PA-18 Supercub variant, the PFD works after Autostart.
For me, same issue with the cache: I have to have it deleted it or to "clean" it from 
```
        <!-- <knob-mode type="int">4</knob-mode>
        <ident-btn type="bool">false</ident-btn> -->
```
On the PA-18 Supercub, the PFD temporarily switches OFF during Autostart, but it works again after. At least the main functions (quick tests).

**Comment by wlbragg** [comment]
@dany93
Finally, I think I figured it out. It appears that it was a loading order, timing issue. I changed the order of when the electrical loads VS the fg1000 and I also delayed their loading with an fdm-initialized timer. [PR](https://github.com/c172p-team/c172p/pull/1427) is ready to merge.

**Comment by dany93** [comment]
Bravo! You got it :smiley:
The c172 FG1000 kap and gfc  work, no longer cache issue after several launches.
Merged. https://github.com/c172p-team/c172p/pull/1427#issue-1817654807

**Comment by dany93** [comment]
On the J3Cub FG1000 and the PA-18 Supercub FG1000 the PFD works too, with no longer cache issue.
Fixed.

Thank you for your searching.
